Don't do string lookups in symex main loop

karkhaz opened the pull request:
There's no need to do a string lookup every time we go around the `goto_symext` main loop, especially for values that don't change. This PR removes `goto_symext`'s ownership of an `optionst` object, and sets the looked-up values as instance members from the `optionst` reference passed during `goto_symext`'s construction.

tautschnig reviewed (changes requested):
I don't think the assertion made holds: the `--depth` command-line option does set this (something I have actually used this morning myself...). What should be done for sure, though, is only reading that option once.

karkhaz commented:
Oops. Thanks.

However, I'm totally confused: I actually don't understand where this is being set. The only line of code that ever does `options.set_option("depth"` is in `cbmc_parse_options.cpp`, and that's a different `options` object than the one that `goto_symext` owns. How does the `optionst` object that `goto_symext` owns ever get the options that are set from the front-end? Sorry if this is a stupid question, but I can't see it...

karkhaz commented:
Never mind, [it was a stupid question](https://github.com/diffblue/cbmc/blob/develop/src/cbmc/bmc.cpp#L308). Will update at some point.

tautschnig commented:
My apologies, I had put answering this on my stack and just not done it yet; really this should be turned into a PR that passes a reference to a `const optionst &` to `goto_symext` - `goto_symext` should neither hold a public member, nor should there be a copy-by-value.

karkhaz commented:
Will do. In fact, goto_symext actually [adds two values](https://github.com/diffblue/cbmc/blob/develop/src/goto-symex/goto_symex.h#L68) to its own `options` in the constructor, which presumably get clobbered by that copy-by-value by `bmct`.

Is it the case that `goto_symext` always needs `"simplify"` and `"assertions"` set to `true` in the `options` object? If so, then I should also set them to true in `bmct`'s own `options` object that will be passed to `goto_symext` by reference, right?

karkhaz committed:
Remove optionst member from goto_symext

goto_symext now has a const reference rather than owning a member. There
is no need to copy an optionst by value from bmct.

karkhaz committed:
Remove string lookups from goto_symext main loop

Lookups into goto_symext's options member have been replaced by accesses
to instance members that are initialized from the optionst reference by
goto_symext's constructor.

This commit fixes #2050.

karkhaz commented:
Just pushed two commits, one which replaces the owned member by a const reference, and one which removes the string lookups from the symex loop.

Note that I've [deleted the setting of "simplify" and "assertions"](https://github.com/diffblue/cbmc/pull/2262/commits/79ad7943c8bfc01bf41921a256634ffd6e448c98#diff-e21df0021b5ea8a478fbdc29faec2399L69) in the `options` object, since it's now a const reference everywhere. This seems safe to me, since those options were immediately clobbered when we did the copy-by-value, but I'll wait to see if the tests agree.

tautschnig commented on src/goto-instrument/accelerate/scratch_program.h:
```diff
@@ -43,7 +43,8 @@ class scratch_programt:public goto_programt
       path_storage(),
-      symex(mh, symbol_table, equation, path_storage),
+      options(),
+      symex(mh, symbol_table, equation, options, path_storage),
```
Please add
```
options.set_option("simplify", true);
options.set_option("assertions", true);
```
in the constructor body.

tautschnig commented on src/goto-symex/goto_symex.h:
```diff
@@ -56,6 +56,8 @@ class goto_symext
     : should_pause_symex(false),
       options(options),
+      max_depth(options.get_unsigned_int_option("depth")),
+      doing_path_exploration(options.is_set("paths")),
```
Would you argue that this fixes #2050? If so, please add it to the commit message.

tautschnig commented on src/goto-symex/goto_symex.h:
```diff
@@ -196,6 +198,9 @@ class goto_symext
   const optionst &options;
 
+  const unsigned max_depth;
+  bool doing_path_exploration;
```
Should this be const (and the const be removed when the partial-merge goes in)?

karkhaz commented:
Thanks, all tests are passing. I did do as you asked in `accelerate/scratch_program.h`, but GitHub isn't folding it.

smowton commented:
Should the PR description be updated?

karkhaz commented:
@smowton done

smowton commented on src/goto-instrument/accelerate/scratch_program.h:
```diff
@@ -43,12 +43,15 @@ class scratch_programt:public goto_programt
       z3(ns, "accelerate", "", "", smt2_dect::solvert::Z3),
       checker(&z3) // checker(&satchecker)
   {
+    options.set_option("simplify", true);
```
Comment *why* these are being unconditionally set, regardless of the user's selections?

smowton replied:
Also suggest commenting that these are local alterations (only applicable to this program), since this sort of code in cbmc is often making global option changes.

karkhaz replied:
@smowton I don't know much about the acceleration code, so I have no idea why these are always set...perhaps @tautschnig knows, or knows who knows? I guess this is Matt Lewis' code, but the migration to GitHub happened after Matt left so I can't ping him on here

tautschnig replied:
That's just me wanting to preserve the prior behaviour, where `goto_symext` would unconditionally set them, which it now no longer does (and therefore they have to be set here). That said, running without simplification turned on would likely yield poor performance. The `"assertions"` option could be scratched, it has no effect here (or in `goto_symext`, it only matters for `goto_check`).

tautschnig replied:
With the above said, @karkhaz could you change this to only set the `"simplify"` option, and add a comment that that's for performance reasons?

karkhaz replied:
@tautschnig will do so. However, unless I'm mistaken, the `goto_symext` constructed by `bmct` does +not+ get the `"simplify"` option switched on (or the `"assertions"` option). Does this also matter for performance, or does it really only matter for the acceleration code?

If it does matter for `bmct`, then I will have to make `bmct`'s option reference non-const, so that I can set `"simplify"` to `true`. I would do this in the `bmct::setup()` method, I suppose. Please let me know if I should do this.

tautschnig replied:
I'd say cbmc_parse_optionst::set_default_options does take care of those?

karkhaz replied:
Amazing that I didn't realise that, given that I wrote that method. Thanks.

smowton commented:
(also the PR text (currently "Nobody ever sets that option, so the conditional won't be entered."))

smowton reviewed (approved):
Thanks! Approving modulo @tautschnig answering the query above

karkhaz commented:
Addressed all suggestions; all tests are passing. Thanks!

peterschrammel commented on src/goto-symex/symex_main.cpp:
```diff
@@ -308,16 +308,13 @@ void goto_symext::symex_step(
-      state.guard.add(false_exprt());
-    state.depth++;
-  }
+  if(max_depth != 0 && state.depth > max_depth)
```
(note for a future PR) This should be changed to use `std::numeric_limits::max` instead of `0`.

tautschnig replied:
#2295.
